### PR TITLE
[SC-116] Jan 10 Spell

### PR DESCRIPTION
**Pull request by hexonaut**
* [Mainnet] Freeze Mom
 * [Mainnet] DAI Hardcoded Oracle
 * [Mainnet] Update wstETH oracle to exchange price (confirm with risk)
 * [Mainnet] Freeze GNO and set LTV to 0
 * [Mainnet] Lido Rewards Program
 * [Mainnet] Pool Implementation Patch

### CODE REVIEW

File: src/proposals/20240110/SparkEthereum_20240110.t.sol [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+    function setUp() public {
+        vm.createSelectFork(getChain('mainnet').rpcUrl, 18934612);  // Jan 4, 2024
+        payload = deployPayload();
+
+        loadPoolContext(poolAddressesProviderRegistry.getAddressesProvidersList()[0]);
+    }
+
+    // --- Configuration Changes ---
+
+    function testFreezerMomDeploy() public {
```
Comment (barrutko) [inline src/proposals/20240110/SparkEthereum_20240110.t.sol:78]: One and the only one style nitpick - I'd rename all `testDescriptionOfTheTest` named tests to `test_descriptionOfTheTest`

**Review by barrutko — COMMENTED** [review]
- `ISparkLendFreezerMom` was copied correctly without any changes
- `setIncentivesController` is correctly called where necessary
- GNO market is correctly frozen and its LTV is correctly set to 0
- Freezer is correctly activated
- Rewards are correctly activated
- DAI oracle is correctly updated (and address points to the right contract)
- wstETH oracle is correctly updated (and address points to the right contract)

LGTM, happy to approve once the polls links are added.

File: src/proposals/20240110/SparkEthereum_20240110.t.sol (hunk 1 above)
Comment (barrutko) [reply src/proposals/20240110/SparkEthereum_20240110.t.sol:78]: Added this in a PR. You can merge it, or close it, your call.
https://github.com/marsfoundation/spark-spells/pull/30

File: src/proposals/20240110/SparkEthereum_20240110.t.sol [hunk 2]
```diff
@@ -0,0 +1,637 @@
...
+    }
+
+    function assertIncentivesController(address asset, address _incentivesController) internal {
+        DataTypes.ReserveData memory reserveData = pool.getReserveData(asset);
+        assertEq(IIncentivizedERC20(reserveData.aTokenAddress).getIncentivesController(),            _incentivesController);
+        assertEq(IIncentivizedERC20(reserveData.variableDebtTokenAddress).getIncentivesController(), _incentivesController);
+        assertEq(IIncentivizedERC20(reserveData.stableDebtTokenAddress).getIncentivesController(),   _incentivesController);
+    }
+
+    function testUpdateRewardsController() public {
```
Comment (lucas-manuel) [inline src/proposals/20240110/SparkEthereum_20240110.t.sol]: Nice

File: src/proposals/20240110/SparkEthereum_20240110.t.sol [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+    function assertFrozen(string memory assetSymbol, bool frozen) internal {
+        assertEq(_findReserveConfigBySymbol(createConfigurationSnapshot('', pool), assetSymbol).isFrozen, frozen);
+    }
+
+    function assertPaused(string memory assetSymbol, bool paused) internal {
+        assertEq(_findReserveConfigBySymbol(createConfigurationSnapshot('', pool), assetSymbol).isPaused, paused);
+    }
+
+    function test_freezerMomE2E() public {
+        GovHelpers.executePayload(vm, payload, executor);
```
Comment (lucas-manuel) [inline src/proposals/20240110/SparkEthereum_20240110.t.sol:567]: NOT BLOCKING DEPLOYMENT - Lets update this test to use all the real deployed freeze/pause spells and give each of them the hat. I would also just make each its own test so we're working against clean state each time.

File: src/proposals/20240110/SparkEthereum_20240110.sol [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    function getIncentivesController() external view returns (address);
+    function setIncentivesController(address controller) external;
+}
+
+/**
+ * @title  January 10, 2024 Spark Ethereum Proposal
+ * @author Phoenix Labs
+ * @dev    Activate Freezer Mom, DAI oracle to hardcoded $1, wstETH oracle assume 1:1 stETH peg, Freeze GNO, Activate Lido Rewards.
+ * Forum:  https://forum.makerdao.com/t/spark-spell-proposed-changes/23298
+ * Polls:  TODO
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/proposals/20240110/SparkEthereum_20240110.sol]: added

File: src/proposals/20240110/SparkEthereum_20240110.t.sol [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+        ReserveConfig[] memory allConfigsAfter = createConfigurationSnapshot('', pool);
+
+        gnoConfigBefore.ltv      = 0;
+        gnoConfigBefore.isFrozen = true;
+        _validateReserveConfig(gnoConfigBefore, allConfigsAfter);
+
+        _validateAssetSourceOnOracle(poolAddressesProvider, DAI,    DAI_ORACLE_NEW);
+        _validateAssetSourceOnOracle(poolAddressesProvider, WSTETH, WSTETH_ORACLE_NEW);
+    }
+
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/proposals/20240110/SparkEthereum_20240110.t.sol:155]: added

File: src/proposals/20240110/SparkEthereum_20240110.t.sol [hunk 6]
```diff
@@ -0,0 +1,637 @@
...
+    function test_gnoDisabledE2E() public {
+        uint256 snapshot = vm.snapshot();
+
+        // User has a dust amount still they can supply
+        vm.startPrank(GNO_USER);
+        pool.supply(GNO, 1, GNO_USER, 0);
+
+        // User can borrow more DAI
+        pool.borrow(DAI, 1, 2, 0, GNO_USER);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/proposals/20240110/SparkEthereum_20240110.t.sol:609]: done

**Comment by lucas-manuel** [comment]
Approving for deployment after doing the following:

- Checked forum link
- Checked all poll links
- Checked all addresses
- Made repo to check bytecode of contracts against local deployments: https://github.com/marsfoundation/bytecode-verification. Checked the following:
  - DAI_ORACLE
  - FREEZER_MOM
  - TRANSFER_STRATEGY
- Reviewed all functions in spell as well as their corresponding tests
- Manually checked that all the reserves are covered in `testUpdateRewardsController`
- Manually checked deployer is NOT the owner of the freezer mom, the spark proxy is
- Reviewed all end to end tests

LGTM, good to merge once deployed spells are used in the `test_freezerMomE2E` test and have their bytecode verified.

**Review by lucas-manuel — DISMISSED** [review]
Verified spell bytecodes, LGTM

**Review by barrutko — APPROVED** [review]

**Review by lucas-manuel — APPROVED** [review]
Approving updated changes after doing the following:

- Reviewed diffs from AAVE contracts on mainnet https://www.diffchecker.com/8Gmr4iAA/ https://www.diffchecker.com/vxJ6Zafn/
- Created working POC to further understand implications of patch https://github.com/marsfoundation/sparklend-incident-response/tree/2024-01-10-poc
- Reviewed diff of patch against deployed v1.18.0 version https://github.com/aave/aave-v3-core/compare/v1.18.0...marsfoundation:aave-v3-core:jan-10-patch
- Verified that the deployed contracts matched the reviewed diff on both mainnet and gnosis chain by compiling the contracts and verifying them against the deployed addresses using sourcify
- Reviewed the updated spell changes
- Checked all new addresses
- Reviewed the updated spell tests

LGTM

File: src/proposals/20240110/SparkEthereum_20240110.t.sol (hunk 3 above)
Comment (hexonaut) [reply src/proposals/20240110/SparkEthereum_20240110.t.sol:643]: this is done now btw can close

**Review by barrutko — DISMISSED** [review]

**Review by lucas-manuel — APPROVED** [review]
LGTM with deployed addresses

**Review by lucas-manuel — APPROVED** [review]

**Review by barrutko — APPROVED** [review]

**Pull request merged**